design/014-secretstore-generator-v2.md

```diff
@@ -0,0 +1,198 @@
+```yaml
+---
+title: SecretStores and Generators v2
+version: v1alpha1
+authors: Gustavo Carvalho
+creation-date: 2025-05-18
+status: approved
+---
+```
+
+# SecretStores and Generators v2
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+This Document describes a design proposal for SecretStores and Generators on its v2 version.
+
+## Motivation
+
+Currently, There are a number of features that cannot be easily done with the current codebase / CRD structure for SecretStores and Generators, including:
+
+* Different Provider Versioning;
+* Allowing first-class support for out-of-tree providers;
+* Allowing to install/uninstall SecretStores and Generators wanted / not wanted by the user.
+* Referent authentication modes are implicit and hard to learn / use.
+
+Furthermore, Generators and SecretStores themselves have different core features, while in reality, we should make them have the same feature set, if not sharing the same codebase:
+
+* SecretStores support ClusterSecretStores with Referrent Authentication and cross namespace generation; Generators dont;
+* Generators support GeneratorState to track down eventual cache information on the CRD; SecretStores dont;
+* SecretStores Are periodically refreshed; Generators arent;
+* Badly configured SecretStores prevent jobs to happen via a gating mechanism; Generators dont;
+
+## Proposal
+
+The proposal is a new CRD for both SecretStores and Generators that unifies all of their feature sets, bringing the best of the worlds. Since this is a significant change, the idea is to also leverage a new controller and client manager for this CRD, in order to make it easier to maintain and scale.
+
+While changing the provider interface for both SecretStores and Generators itself is a non goal, it might be needed to accomodate this proposal.
+
+### SecretStore
+SecretStore manifests looks very much like our original implementation. Provider configuration resides on `spec.provider` field. 
+```yaml
+apiVersion: secretstore.external-secrets.io/v2alpha1
+kind: AWSSecretManager
+metadata:
+  name: my-aws-store
+  namespace: default
+spec:
+  refreshInterval: 1m
+  controller: dev
+  retrySettings:
+    maxRetries: 3
+    retryInterval: 10s
+  providerConfig:
+    type: InTree/OutOfTree
+    address: http+unix:///path/to/socket.sock
+  cachingPolicy: ProviderDefined/Secrets/Auth/All/None
+  cacheConfig: # if Caching Policy is Secrets/Auth/All 
+      ttl: 1m
+      maxEntries: 100
```

Comment:
This would allow for implementations like https://github.com/external-secrets/external-secrets/pull/4727 to be completely safe and make sure Secret Managers themselves can be free to implement this sort of protection.

CC @watsonian